### ISSUE TITLE
Feedback Requested: Collectd Release Process Proposal

### ISSUE THREAD

**Issue by sunkuranganath** [collectd/collectd#3223]
Hello Collectd’ers,

In order to bring together a structure and known cadence to Collectd releases, @mrunge , @tokkee and I have been working together to bring forward to the community a release process proposal.

The idea is to share the burden of releasing, tagging and maintaining releases and establish few guidelines for community ownership.
We are looking for feedback and progress towards arriving at a common ground. Appreciate your help.

Please go through following proposal shared via Google docs link [1]. 

[1] https://docs.google.com/document/d/1Ha9zPyKa_1pBar58LUPRf4EV2ap1sMcrOtRz3qauogo/edit?usp=sharing

**Comment by rpv-tomsk** [comment]
Collectd has no active developers, has no active maintainers, has no publicaly active Owner. face it.
Team did all possible to terminate development, while you  try to do the opposite.  I think, your try will have no success (you simply has no appropriate permissions on GitHub project), don't waste your time, guys.

I think, you will have nothing good to release in nearest, simply because nobody is interested in PR's merging, no matter how useful the changes.  Only destroying changes accepted freely.

**Comment by sunkuranganath** [comment]
Please check https://github.com/collectd/collectd/pull/3224

**Comment by zebity** [comment]
Hi collectd'ers,

agree with core objective of seeing to get "know cadence" going on collectd.

Here are my thoughts and outline of issues which I am seeing on this:

1. Need to define clearly (or simply restate [versioning](https://en.wikipedia.org/wiki/Software_versioning) principles are. To my mind it should be something like (taking 5.9.2 as example)>
Major (ie 5) - reflects change to core collectd driver code
Minor (ie 9) - reflects additions / removals if "significant" (and given passions and effort put into plug-ins they are likely to be all considered significant)
Patch (ie 2) - reflects fixes, plug-in enhancements, beautification and other maintenance activities

2. Need to ensure that CI process defines clear test objectives, currently this process appears to be focussed on testing that things compile, rather than run.
It should be that a bump in Major, Minor needs to be functionally verified across range of distribution (see point 3), while patch release can go through a more limited set of validations

3. Need to define they set of targets that must be covered for Major / Minor release.
My observation is that there are 3 primary platforms for which there is plug-in code for:
Linux variations, BSD, Solaris and Windows. I believe that release type requires clear definition of verification quality steps. Here is possible set:
Major - must be functionally verified on each of major OS: Linux, BSD, Solaris, Windows (feel free to comment to this list ;-) )
Minor - two cases:
    OS specific Plugin - verify against variants (ie):
       BSD - FreeBSD, OpenBSD, NetBSD etc
       Linux - Fedora / Redhat (RPM Packaging), Debian / Ubuntu (debian packaging), others ?
       Solaris - is there more than one ?
       Windows - leave this to Windows experts to define
Patches - can be covered by exiting CI process (but see item 4 below)

4. CI process should cover more of target release platforms, we should extend this to cover: BSD variant, windows variant & RH / Fedora & Debian / Ubuntu variant

5. What is the minimal set of critical components that should be part of functional smoke test. My nominations are: collectd (deamon duuuh), virt (essential for KVM based infrastructure including oVirt and OpenStack), Kafka, Prometheus, RRD, cpu, memory (base ones enabled by default config), log_writer , others ???

5. Once we have got to here and with right approval (based on impact / scope) the release process should automatically flow through to all the downstream distribution processes.


Now some observations on issues:

1. There appears to be issue with "trust" (see https://github.com/collectd/collectd/pull/3318) that @rpv-tomsk keeps raising - plug-in developers vs. maintainers. None of this make sense and only leads to problems... code is code and once released into community the notion of "ownership" is a bit tenuous anyway. To me this is just a matter of courtesy rather rules. We should just accept this and move on. In fact to me Open Source is a bit like raising children, you to your best and then they are released to the wild and must fend for themselves without continual parental protection ;-)

2. There is issues with getting offical approvals and flow through to down stream repositories (collectd.org, debian.org etc) - @mrunge Matthias has it hit these with both 5.9 and now 5.10 cuts and I am hitting with my debian struggles (https://github.com/collectd/pkg-debian/pull/18).

3. There are issues with responsiveness that are frustrating people: see https://github.com/collectd/pkg-debian/issues/17 & https://github.com/collectd/collectd/pull/3273 as examples but there are many others

Finally, why is this important and who are key consumers of collectd:

Just because I think it is always important to keep audience in mind, I believe that collectd is valuable endevour and useful for the following:
1. OpenStack, oVirt Instrumentation
2. Ceph monitoring (part of above)

How do we address issues: Simple open issue via GitHub and act transparently to agree and action resolution.

Alternatives:
Baby spits the dummy ... :-) 
Shift efforts to Fluentd or alternate collection platform is we cannot resolve via collectd community process.

People who appear to be involved, contributing and interested.
Here is my quick list of people who appear be to involved and should help drive process:

@rpv-tomsk - a major contributor to keeping code running
@mrunge - really helpful in keeping releases driving along
@sunkuranganath - raised this issue in interest of community
@rubenk - another current active participant
@octo - the parent with noisy children ;-)
@tokkee - another parent with noisy children
@mfournier - another parent with noisy children

**Comment by mrunge** [comment]
Thank you.

all valid points.

The word "trust" pops up here and there, and may be misleading here. I believe, if comes from the group here in github named "trusted contributors". They were the original maintainers with write permission to the repository, i.e even the ability to merge changes without a review.

There is clearly a lack of code reviews, and @octo proposed to introduce/use the code-ownership model github offers. Some plugins require some deep knowledge of the plugin itself and we decided to try out allowing "plugin owners" to approve patches for these plugins, without being a "trusted contributor" (in the sense described above).

Introducing [code owners](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) now had the side-effect, that original maintainers are not able to merge code without review anymore. That probably lead @rpv-tomsk think, he's not trusted anymore.

I must admit, it is clearly frustrating trying to dance around issues here, cause by not being able to fix things directly due to lack of permissions. However, I'm not at the point, where I'd be looking at alternatives.

**Comment by rpv-tomsk** [comment]
>it comes from the group here in github named "trusted contributors". They were the original maintainers with write permission to the repository, i.e even the ability to merge changes without a review.
>There is clearly a lack of code reviews, and..

Yeah, "trusted contributors" has permission to merge changes without review. Now it has not.
How this helps to get MORE code reviewers?

Again, cases:

I) PR from developers

1) Developer creates PR. PR pass all CI checks.
2) Maintainer merges PR into trunk.
3) Trunk has another CI checks, and some check fails.

Maintainer has no option to solve occured issue w/o other maintainer review.
And there is problem with reviews, so he can wait for months, while he is responsible for merging PR which breaks CI in trunk.

That is OK ? I think, No.

II) Maintainer fix in plugin with "code owners"

If maintainer (note: member of Collectd Team) wants to make changes in plugin, covered by Code Owners, these changes MUST be approved by "code owner" (another maintainer approve is not enough). 

Who is more important for project - member of Collectd Team or person from third party company?

III) Maintainer's minor changes (similar to I)

While my active work on Collectd, I did many minor changes and merge it w/o reviews. 
Was there problems due to such my activity? There was no.

Personally, I don't like to wait for months while somebody pay attention to PR.

**Comment by kkepka** [comment]
Thank you all for comments/feedback above.

Having predictable release cadence is a good input for aligning with it.

Thanks to Matthias for keep releases appearing, much appreciated. However it's not great put that pressure only on single person, rather better would be to share this effort across more heads. That's why idea of this more distributed release process appeared so more people could keep those gears moving. Doc is just a proposal place to agree on terminology and steps.

I really like the idea of extending CI with functional validation, that would put more confidence and visibility on what is actually working and how.
Starting with sanity/smoke tests for smaller amount of features and then expand sounds like a good first steps in enabling this.
What do you think about moving towards kind of continuous delivery of potential release candidate packages in the future?
Other ideas would be like distributed CI feedback from multiple parties (e.g. for less accessible hw/sw parts), test environment preparation playbooks for dependencies, etc.
In general automation usually helps in preparing releases.

I think we all appreciate Pavel's and other contributors work in keeping code running and valuable feedback on issues and code reviews - thank you Pavel and all contributors. I believe code owners feature was enabled in good faith to resolve other issue and introduced side-effect was not the intention. This part still requires some effort on removing roadblocks from both maintainers and contributors.

Acting with full transparency is a good direction, ++ here.

Just a few thoughts.

**Comment by rpv-tomsk** [comment]
>I believe code owners feature was enabled in good faith to resolve other issue and introduced side-effect was not the intention.

... but there is no changes in this feature implementation.

From my POV, when new serious changes introduced, responsible person should be ready to get feedback and made operational changes. By the fact, we get tightened rules only, and no any responses.

**Comment by zebity** [comment]
HI @mrunge , thanks for clarify "trust" concern. No point in tilting at windmills.

@rpv-tomsk like @mrunge (and myself by misfortunate nessessity) I agree we have to allow for critical fixes to be done on plug-ins, without owner approval, if there is agreed issue that must be resolved. The typical way these things are handled is via priority flagging. If there is a critical defect in a plugin and known way to fix it then waiting for months for approval does not make sense. There needs to be the ability to override the approval process to get fix in ASAP. For feature updates, the time from proposal to merge can be much longer.

On extending CI tooling, I am pretty proficient with BSD variants and could help here. The low $$ but likely prone to failure option would be to set of a test harness machine on my own network, other option would be see if we could make it small enough to run on AWS for free. If we do extend CI then as per @kkepka , this would need to be in some distributed & redundant implementation to allow for the fact that individual testing machines are not available to avoid stalling the build pipeline.

Cheers,

Zebity.

**Comment by rpv-tomsk** [comment]
>I agree we have to allow for critical fixes to be done on plug-ins, without owner approval, if there is agreed issue that must be resolved.

Not only critical fixes, but any others too.
Look at the behavior of people (#3331 as fresh example): Nobody wants to merge others pull requests. The same applied to reviews. The same applied to codeowners people from "BIG COMPANIES" - they are not interested in NIH changes and may be has no time to review them at all. 

And while somebody wants to do some work for project, he get administrative/technical barriers, even he became part of Team.

**Comment by bzed** [comment]
There seems to be yet another bug in the release process: The tarball on collectd.org does not contain the same source as what is taged as 5.9.2 here... (and I'm not talking about configure and friends).

**Comment by mrunge** [comment]
> There seems to be yet another bug in the release process: The tarball on collectd.org does not contain the same source as what is taged as 5.9.2 here... (and I'm not talking about configure and friends).

What do you mean by "another" bug? There are two reported
- https://github.com/collectd/collectd/issues/3326
- https://github.com/collectd/collectd/issues/3293

Do you see more?

I am expecting collectd-5.10 any day now, so that should be fixed then.

**Comment by bzed** [comment]
Yes, as mentioned, the tarball does not contain what was tagged on github.
https://salsa.debian.org/debian/pkg-collectd/commit/95063e7d49a3a19c21a8ba3c9972bfd8f20c15cb is the diff - and if you ignore configure and others, the result is still very confusing. Also the tarball on collectd.org contains bugs, that are fixed in the github tag.
